074-connector-offsets-support.md

```diff
@@ -0,0 +1,226 @@
+
+The ConfigMap the operator creates will be called `{connector}-offsets`, where `{connector}` is the name of the KafkaConnector CR that was annotated.
+The ConfigMap will contain a single file called `offsets.json` that contains the response received from the GET /connectors/{connector}/offsets endpoint.
+The KafkaConnector CRD will also be updated to contain a new property called `listOffsetsCM` that when set will override the name of the ConfigMap the operator creates/updates with the endpoint response.
+
+For example for a source connector the ConfigMap might look like:
+
+```yaml
+apiVersion: v1
+kind: ConfigMap
+metadata:
+  name: my-source-connector-offsets
+data:
+  offsets.json: |
+    {
+      "offsets": [
+        {
+          "partition": {
+            // Connector-defined source partition
+          },
+          "offset": {
+             // Connector-defined source offset
+          }
+        }
+      ]
+    }
+```
+
+For a sink connector it might look like:
+
+```yaml
+apiVersion: v1
+kind: ConfigMap
+metadata:
+  name: my-sink-connector-offsets
+data:
+  offsets.json: |
+    {
+      "offsets": [
+        {
+          "partition": {
+            "kafka_topic": "my-topic"
+              "kafka_partition": 2
+          },
+          "offset": {
+            "kafka_offset": 4
+          }
+        }
+      ]
+    }
+```
+
+### Altering offsets
+
+To alter offsets the user will annotate the KafkaConnector resource with the new annotation `strimzi.io/connectorOffsets` set to `alter`.
+After the annotation is added, on the next reconciliation the operator will call the PATCH /connectors/{connector}/offsets endpoint to alter the offsets.
+Once the patch is complete the operator will then remove the annotation from the KafkaConnector CR.
+
+By default, the operator will construct the request body from a ConfigMap called `{connector}-offsets`, where `{connector}` is the name of the KafkaConnector CR that was annotated.
+This matches the name used in the list operation, so that a user can list offsets, then edit the ConfigMap the operator created, then request the offsets be altered using that same ConfigMap.
+
+The KafkaConnector CRD will also be updated to contain a new property called `alterOffsetsCM` that when set will override the name of the ConfigMap the operator uses to populate the request body.
+
+Notes:
+* The format of the request body will be validated to make sure it is valid JSON, but the operator won't perform any further validation, instead relying on the API endpoint to return a reasonable error message.
+* If the request fails the KafkaConnector CR status will be updated to include a warning message with the response from the API, e.g.
+  ```yaml
+  - lastTransitionTime: "2024-06-04T08:44:15.913138115Z"
+    message: Failed to alter the connector offsets due to "message from endpoint".
+    reason: AlterOffsets
+    status: "True"
+    type: Warning
+  ```
```

Comment:
I don't think we should retry unless the user explicitly re-annotates the resource, since there aren't any expected failures that should resolve automatically, however I worry that expecting the user to read the log to spot a failure will result in people not noticing it hasn't worked.

Is it feasible to leave the condition there for a period of time? Or could we use another annotation to communicate that a failure ocurred?

Reply:
That is not how the logic works in that case:
1) You remove the annotation or update the status to add the condition
2) That triggers reconciliation
3) This reconciliation doesn't do the operation anymore as the annotation is not there
4) The condition is removed as it is not valid anymore

Doing some time test would IMHO add a lot of complexity to the operator. I think we need either some other way to do this or we should stick with log messages only.

Or we should keep the annotation and retry it. I agree that it seems a bit silly and the retrying alone might not solve much. But maybe it helps the user to realize it doesn't work?

Reply:
Would it work to add the condition but also keep the annotation? That way the first time the annotation is added a new reconciliation would be triggered, however since the error message returned should be identical the second attempt would not change the status of the KafkaConnector. So then after that the reconciliation would just happen every 2 minutes, giving the user time to notice the condition. What do you think @scholzj?